### PR TITLE
RFC: metal-stack support

### PR DESCRIPTION
**Pull request by majst01**
**What this PR does / why we need it**:
With this PR we gain metal-stack, bare-metal as a service with metal-stack.

### CODE REVIEW

**Comment by hardikdr** [comment]
@majst01 Thanks a lot for the PR. 
We shall take a look soon :)

File: go.sum [hunk 1]
```diff
@@ -67,21 +76,110 @@ github.com/evanphx/json-patch v4.2.0+incompatible/go.mod h1:50XU6AFN0ol/bzJsmQLi
 github.com/fsnotify/fsnotify v1.4.7 h1:IXs+QLmnXW2CcXuY+8Mzv/fWEsPGWxqefPtCP5CnV9I=
 github.com/fsnotify/fsnotify v1.4.7/go.mod h1:jwhsz4b93w/PPRr/qN1Yymfu8t87LnFCMoQvtojpjFo=
 github.com/ghodss/yaml v0.0.0-20150909031657-73d445a93680/go.mod h1:4dBDuWmgqj2HViK6kFavaiC9ZROes6MMH2rRYeMEF04=
+github.com/globalsign/mgo v0.0.0-20180905125535-1ca0a4f7cbcb/go.mod h1:xkRDCp4j0OGD1HRkm4kmhM+pmpv3AKq5SU7GMg4oO/Q=
```
[the remark this answers is not in the document]
Comment (majst01) [reply go.sum]: I guess this is a indirect dependency, will check and report back

File: kubernetes/deployment/deployment.yaml [hunk 2]
```diff
@@ -14,7 +14,14 @@ spec:
       containers:
       - name: machine-controller-manager
         image: eu.gcr.io/gardener-project/gardener/machine-controller-manager:0.5.0
-        imagePullPolicy: Always
+        imagePullPolicy: Allways
```
[the remark this answers is not in the document]
Comment (majst01) [reply kubernetes/deployment/deployment.yaml]: uups

File: Dockerfile [hunk 3]
```diff
@@ -4,16 +4,14 @@ FROM golang:1.13.5 AS builder
...
+RUN .ci/build \
+ && strip /go/src/github.com/gardener/machine-controller-manager/bin/rel/machine-controller-manager
 
 #############      base                                     #############
 FROM alpine:3.11.2 as base
 
 RUN apk add --update bash curl tzdata
 WORKDIR /
 
-#############      machine-controller-manager               #############
```
[the remark this answers is not in the document]
Comment (afritzler) [reply Dockerfile]: That was actually not necessary.  `FROM base AS machine-controller-manager` must be added again to reduce the image size.

[the remark this answers is not in the document]
Comment (majst01) [reply Dockerfile]: Adding this additional FROM stanza does not change the resulting image size:

```

# Without FROM base as machine-controller-manager
eu.gcr.io/gardener-project/gardener/machine-controller-manager            v0.27.0-dev                                2627a567138b        18 seconds ago      73.5MB
# With FROM base as machine-controller-manager
eu.gcr.io/gardener-project/gardener/machine-controller-manager            v0.27.0-dev                                2627a567138b        3 minutes ago       73.5MB
```

File: METAL.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Metal usage
```
[the remark this answers is not in the document]
Comment (majst01) [reply METAL.md]: done

File: Makefile [hunk 5]
```diff
@@ -74,13 +74,13 @@ docker-login:
 
 .PHONY: docker-push
 docker-push:
-	@if ! docker images $(IMAGE_REPOSITORY) | awk '{ print $$2 }' | grep -q -F $(IMAGE_TAG); then echo "$(IMAGE_REPOSITORY) version $(IMAGE_TAG) is not yet built. Please run 'make docker-images'"; false; fi
+	@if ! docker images $(IMAGE_REPOSITORY) | awk '{ print $$2 }' | grep -q -F $(IMAGE_TAG); then echo "$(IMAGE_REPOSITORY) version $(IMAGE_TAG) is not yet built. Please run 'make docker-image'"; false; fi
```
[the remark this answers is not in the document]
Comment (majst01) [reply Makefile]: oh yes, good catch, fix all.

File: pkg/controller/secret_util.go [hunk 6]
```diff
@@ -54,12 +54,18 @@ func (c *controller) existsMachineClassForSecret(name string) (bool, error) {
...
+		return false, err
+	}
 
 	if len(openStackMachineClasses) == 0 &&
 		len(gcpMachineClasses) == 0 &&
 		len(azureMachineClasses) == 0 &&
 		len(packetMachineClasses) == 0 &&
+		len(metalMachineClasses) == 0 &&
 		len(alicloudMachineClasses) == 0 &&
+		len(metalMachineClasses) == 0 &&
```
[the remark this answers is not in the document]
Comment (majst01) [reply pkg/controller/secret_util.go]: thanks, done

File: pkg/driver/driver_metal.go [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+// GetExisting method is used to get machineID for existing Metal machine
+func (d *MetalDriver) GetExisting() (string, error) {
+	return d.MachineID, nil
+}
+
+// GetVMs returns a machine matching the machineID
+// If machineID is an empty string then it returns all matching instances
+func (d *MetalDriver) GetVMs(machineID string) (VMs, error) {
```
[the remark this answers is not in the document]
Comment (majst01) [reply pkg/driver/driver_metal.go]: good point, will need to check with colleauges

File: cmd/machine-controller-manager-cli/METAL.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# Metal usage
```
Comment (prashanth26) [inline cmd/machine-controller-manager-cli/METAL.md]: This file should be moved to `/docs/development`. And could we also make it generic enough to make instructions generic across different providers?  You could leave a line with the example, for now, to run metal maybe?

File: pkg/driver/driver_metal.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (prashanth26) [reply pkg/driver/driver_metal.go]: @hardikdr - I think it would be recommended considering that we should check for VMs before deleting them

**Comment by prashanth26** [comment]
Hi @majst01 ,

Sincere apologies for the delay in reviewing the PR as we have been having a lot of work currently being done for moving the providers out of tree. Hence we were contemplating how to take in your PR. From our discussions, we would be happy if your changes go out of tree as it would be easier for you to maintain it. Also if you merge it in tree for now, we would eventually like to move it out of tree hence it would be a two step approach. The out of tree should take about a month more which is a negative.

However, if time is an issue with your development we can think of taking up this PR in tree and later migrate it out of tree. 

Regards,
Prashanth

**Comment by majst01** [comment]
Hi @prashanth26 ,

We are not in a hurry, and if, as you say it will take a month or so to be able to keep provider support out of tree, we are fine with that. 

May i keep the PR as a reminder open ?
Greetings
Stefan

**Comment by prashanth26** [comment]
Makes sense. Sure sounds good to me 👍

**Comment by majst01** [comment]
hi @prashanth26,

any estimates when the out of tree support is expected ? Getting nervous because we already lost contact to the current code base, and the longer we wait the worse.

**Comment by prashanth26** [comment]
Hi @majst01 ,

The OOT is already available since last week (and release). Refer - https://github.com/gardener/machine-controller-manager/blob/master/docs/development/cp_support_new.md. I wanted to improve the documentation a little, however, it is still in a decent state. If you follow this doc, hopefully, should work. 

Just heads up: The rename-project and rename-provider scripts might be outdated and might require manual intervention there. But it is hopefully a trivial thing.

**Comment by majst01** [comment]
Had a look at the sample provider, first of all `make rename-project` and `make rename-provider` dont work because the layout of pkg is totally different now. I can see that for example the gcp provider is similar to what is expected in hack/rename-project and hack/rename-provider, but the sample-provider is not.

what is the expected an most actual package structure ?

**Comment by prashanth26** [comment]
>  first of all make rename-project and make rename-provider dont work because the layout of pkg is totally different now.

Yes. This is something I am yet to fix for the `SampleProviders` repo. I will fix it soon and you should be able to get around it until then. 

> what is the expected an most actual package structure ?

There is no fixed package structure. However, the sampleProvider should provide you a bluePrint. You can refer to these implementations for further implementation doubts. 

1. https://github.com/prashanth26/machine-controller-manager-provider-aws/tree/spliting-controllers2
2. https://github.com/gardener/machine-controller-manager-provider-vsphere
3. https://github.com/gardener/machine-controller-manager-provider-gcp

**Comment by prashanth26** [comment]
Also, this should help you with the expected error code handling - https://github.com/gardener/machine-controller-manager/blob/master/docs/development/machine_error_codes.md.

**Comment by majst01** [comment]
@prashanth26 thanks for the pointers

**Comment by majst01** [comment]
Closing, we have a OOT version now in place.

**Pull request closed without merging**